File: Java/ROUND_3_B/src/main/java/Sorter.java


public abstract class Sorter {

	public static int[] sort(SortKind kind, int[] input) {
		switch (kind) {
			case BUBBLE:
				return new Bubble().doSort(input);
			case QUICK:
				return new Quick().doSort(input);
			case INSERTION:
				return new Insertion().doSort(input);

		}
		throw new RuntimeException("unknown sort kind " + kind);
	}

	protected abstract int[] doSort(int[] input);

	protected void swap(int[] input, int index1, int index2) {
		int first = input[index1];
		int second = input[index2];
		input[index1] = second;
		input[index2] = first;
	}


}

File: csharp/ROUND_3/ROUND_3_B/src/Sorter.cs
namespace ROUND_3_B;

public abstract class Sorter
{
    public static int[] Sort(SortKind kind, int[] input)
    {
        return kind switch
        {
            SortKind.BUBBLE => new Bubble().DoSort(input),
            SortKind.QUICK => new Quick().DoSort(input),
            SortKind.INSERTION => new Insertion().DoSort(input),
            _ => throw new ArgumentException($"Unknown sort kind '{kind}'", nameof(kind))
        };
    }

    protected abstract int[] DoSort(int[] input);

    protected void Swap(int[] input, int index1, int index2) {
        int first = input[index1];
        int second = input[index2];
        input[index1] = second;
        input[index2] = first;
    }
}